### DISCUSSION TITLE
Partially execute a parse action

### DISCUSSION THREAD

**Discussion by sschneider-ihre-pvs** [colinhacks/zod#960] [Q&A] [answered]
So using vee-validate with zod for form validation brought up a thing with how zod is used in that context. So lets take a z.object({}) with some validation rules for fields. Since there is only .parse or .safeParse and so on for the whole object, this is always applied when the form validates no matter if all those values have been changed.
This brought up a problem with legacy data that might be not valid but also not changeable.
So my question is, is it possible to define a z.object({}) but then only parse certain elements of it like z.object({a, b}).parse(data, ['a']) something like that, so instead of form validation, field validation would be possible if there was a way to partially executing a schema.
One obstacle would remain though, handling refines that have a reference to a field that is not selected have to be handled too.

**Comment by scotttrinh** [comment]
That's definitely achievable with `.partial`, but as you noted, refinements make that a bit harder. I suggested an approach on a similar issue [here](https://github.com/colinhacks/zod/issues/479#issuecomment-854822386) which embraces the potential `Partial` nature so you can reuse the same refinement function on a partial or full schema.

If you really only care about validating certain data, but don't want the other fields to be stripped from the object, you could use `passthrough` and only define the fields you care about on the schema. It would only check those fields, but the parsed data would contain any other fields which might have been defined. You have to be careful with that since TypeScript won't know anything about those other properties, but if you just need to pass it back to your backend, that _should_ be fine. If you _do_ need to use those fields, I would suggest not lying to the compiler and properly handle those "optional" fields even if you don't _want_ them to be optional.

**Comment by cefn** [comment, marked as answer]
Have you considered the approach below, that allows you to base a new schema on picking any parts of an existing schema that you want...

```typescript
import {z} from "zod";

const tightSchema = z.object({
  id:z.string(),
  rev:z.string(),
  message:z.string(),
})

const {shape} = tightSchema;

const looseSchema = z.object({
  message:shape.message,
  id: shape.id.optional(),
})
```

![image](https://user-images.githubusercontent.com/159819/149670645-047dfd42-ae18-4842-a205-dc2fce11b39a.png)

**Comment by sschneider-ihre-pvs** [comment]
Thanks for the two suggestions, I will look into it.
